### ISSUE TITLE
Proposal: Sequence Expressions

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#377]
@gafter commented on [Tue Oct 20 2015](https://github.com/dotnet/roslyn/issues/6182)

We propose a form of expression that allows you to declare temporary variables before the final result of the expression. We replace the existing parenthesized expression with the following

> _parenthesized-expression_:
> &nbsp;&nbsp;&nbsp;&nbsp;`(` _sequence-statements_<sub>opt</sub> _expression_ `)`
> _sequence-statements_:
> &nbsp;&nbsp;&nbsp;&nbsp;_sequence-statement_ _sequence-statements_
> &nbsp;&nbsp;&nbsp;&nbsp;_sequence-statement_
> _sequence-statement_:
> &nbsp;&nbsp;&nbsp;&nbsp;_expression-statement_
> &nbsp;&nbsp;&nbsp;&nbsp;_declaration-statement_

The scope of any variable declared in a _parenthesized-expression_ is the body of that _parenthesized-expression_.

The type of a _parenthesized-expression_ is the type of its final expression.

At run-time the statements are executed in sequence, and then the final expression is evaluated, which becomes the value of the whole sequence expression.

The definite assignment and reachability rules need to be specified, but they are trivial.

This is particularly useful in combination with the expression form of the pattern matching construct #5154.

/cc @TyOverby

---

@alrz commented on [Tue Oct 20 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-149676426)

love this :+1:

---

@orthoxerox commented on [Tue Oct 20 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-149713290)

Why not

> parenthesized-expression:
> &nbsp;&nbsp;&nbsp;&nbsp;`{`  sequence-statements<sub>opt</sub> expression-statement `}` 
> sequence-statements:
> &nbsp;&nbsp;&nbsp;&nbsp;sequence-statement sequence-statements
> &nbsp;&nbsp;&nbsp;&nbsp;sequence-statement
> sequence-statement:
> &nbsp;&nbsp;&nbsp;&nbsp;expression-statement
> &nbsp;&nbsp;&nbsp;&nbsp;declaration-statement

?

---

@alrz commented on [Wed Oct 21 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-149714250)

@orthoxerox it will be ambiguous in expression-bodied lambdas (discussed at length in #5402).

---

@gafter commented on [Tue Oct 20 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-149720126)

@orthoxerox Because that would break compatibility by making `(1+1)` a syntax error for two reasons - first, because `1+1` is not one of the expression forms allowed as an _expression-statement_, and second, because it lacks the semicolon that is part of an _expression-statement_.

Oh, wait, you're suggesting that we replace the parens in a parenthesized expression with curly braces? I think that would make lots of things ambiguous.

---

@paulomorgado commented on [Tue Oct 20 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-149743915)

Why is it named sequence?

---

@gafter commented on [Tue Oct 20 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-149744117)

@paulomorgado Because it evaluates the expressions sequentially for side effects. Got a better name?

---

@paulomorgado commented on [Tue Oct 20 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-149752889)

@gafter, this used to be referred as declaration expressions, right.

Aren't you afraid it might be confused with "as opposed to asynchronous expressions"?

And, no! I don't have a better name. Yet!

---

@ufcpp commented on [Tue Oct 20 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-149755288)

What is its result type? That of the last expression? I receive the impression from the name "sequence expression" that its result type is an array of each expression.

---

@gafter commented on [Tue Oct 20 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-149768281)

@paulomorgado No, this is not the same as declaration expressions. That was where you could declare a variable at its first use, something like `M(int x = 12, x*x)`.

The type and value is the type and value of the last expression.

---

@paulomorgado commented on [Wed Oct 21 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-150058483)

See?

---

@gafter commented on [Wed Oct 21 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-150063286)

@paulomorgado I'm not sure you understand. This would not be legal under this proposal:

`M(int x = 12, x*x)`

That's because a local declaration (with or without its semicolon, which is normally part of its syntax) is not an expression.  If you wanted to do something like that with this proposal, you'd have to write

`(int x = 12; M(x, x*x))`

---

@paulomorgado commented on [Wed Oct 21 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-150076727)

OK! Got it, @gafter!

My "See?" comment was regarding this:

@paulomorgado:

> Aren't you afraid it might be confused with "as opposed to asynchronous expressions"?

and this:
@ufcpp 

>  I receive the impression from the name "sequence expression" that its result type is an array of each expression.

But given your last example, **sequence** is starting to make sense to me.

---

@TyOverby commented on [Wed Oct 21 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-150097996)

@gafter Since you invited the bikeshedding, I think "chained expressions" would be a nice name.  It doesn't have the connotation of sounding like a list or a generator, and the "chain" imagery helps show that expressions can depend on results produced by previous ones.

---

@quinmars commented on [Thu Nov 19 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-158234190)

FYI, gcc calls a similar C extension ["statement expression"](https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html).

---

@qrli commented on [Fri Dec 18 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-165733734)

Seems it will introduce a little inconsistency with the for statement, which uses comma.
Ugly code:

``` C#
for (int i = 0, j = 1; (int k = i + j; M(k, k)); i++, j *= 2)
{
}
```

Though people won't mix them in practice.

It will also be possible to write silly program like this, which does not look like C#:

``` C#
public double Foo() => (
   int a = GetSomething();
   Log(a);
   DoSomething(a);
);
```

---

@alrz commented on [Fri Dec 18 2015](https://github.com/dotnet/roslyn/issues/6182#issuecomment-165819721)

@qrli First example could be written like

```
for (int i = 0, j = 1; i + j case var k : M(k, k); i++, j *= 2)
```

And in the second example, last inner semicolon is a syntax error.

```
public double Foo() => (int a = GetSomething(); Log(a); DoSomething(a));
```

By the way, this is just a matter of taste, you should not use expression-bodied methods wherever you _can_, but you might do so if you prefer.

---

@leppie commented on [Thu Mar 24 2016](https://github.com/dotnet/roslyn/issues/6182#issuecomment-200872307)

Sub

---

@alrz commented on [Mon Dec 05 2016](https://github.com/dotnet/roslyn/issues/6182#issuecomment-264870685)

>The scope of any variable declared in a parenthesized-expression is the body of that parenthesized-expression.

This conflicts with current scoping of is var and out var
```cs
var x = (e is T t ? t : foo);
// t is in scope!
```
Probably this should be decided now if this feature is planned for a future release.

**Comment by orthoxerox** [comment]
Dupe of #9? 😀

**Comment by MgSam** [comment]
It looks like the LDM recently [voted down](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-02-21.md) declaration expressions (again). Is this proposal meaningfully different to warrant separate consideration?

**Comment by gafter** [comment]
@orthoxerox No, #9 is a PR which integrated the C# language spec.

@MgSam I don't see any discussion of declaration expressions in those notes. What are you referring to?

**Comment by orthoxerox** [comment]
@gafter sorry, #72

**Comment by MgSam** [comment]
@gafter It's pretty terse but I assumed that was what [this](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-02-21.md#declarations-in-embedded-statements) was referring to.

**Comment by svick** [comment]
@MgSam I *think* that's talking about allowing code like:

```c#
if (condition)
    var foo = bar;
```

**Comment by MgSam** [comment]
@svick But that seems pretty clearly useless.

I thought it was something like this:
``` C#
var a = foo() ? var b = 5; b + 37 : 0;
```
Which seems like it also fits the definition of sequence expression, no?

**Comment by svick** [comment]
@MgSam That's not what [embedded statement](https://github.com/dotnet/csharplang/blob/82c1088104ead6f49db157c77d5acb0c9e3e3bc0/spec/statements.md) means.

**Comment by MgSam** [comment]
Quite right. My mistake.

**Comment by Richiban** [comment]
I like this idea. But, why not use braces and `return` for the syntax? Currently lambdas have two different syntaxes depending on whether they're statement-bodied or expression-bodied.

An expression-bodied lambda:

```csharp
person => person.Name
```

If we want to execute some statements before returning the name, we can write a statement-bodied lambda:
```csharp
person => { Log("Accessing name"); return person.Name; }
```

If we adopt the parens-based syntax for this proposal then example 2 will have an alternative syntax that accomplishes the exact same thing:

```csharp
person => ( Log("Accessing name"); person.Name )
```

Which will be confusing, in my opinion. Wouldn't it instead be better to make a code block evaluate to an expression?

```csharp
var name = {
    Log("Acessing name");
    return person.Name;
}
```

This would then unify the two syntaxes for lambdas. Instead of:

> arguments_list '=>' expression
> | arguments_list '=>' statement_block

only 

> arguments_list '=>' expression

would exist.

Also, with the proposed syntax, expression-bodied members mean that there are two ways to write any method:

```csharp
public string DoStuff()
{
	Log("Doing stuff");

	return "some result";
}
```

and

```csharp
public string DoStuff() => (
	Log("Doing stuff");

	"some result");
```

Again, with blocks-as-expressions the only thing that changes is the presence of the `=>` symbol:

```csharp
public string DoStuff() =>
{
	Log("Doing stuff");

	return "some result";
}
```

I think this results in a more consistent syntax overall.

**Comment by HaloFour** [comment]
@Richiban 

That would conflict with existing block syntax since the expression is not at all dependent on an assignment.

**Comment by orthoxerox** [comment]
@Richiban you could *probably* rewrite the grammar to make blocks expressions, but `return` is out of the question, since it always returns from the method, not the innermost block.

**Comment by Richiban** [comment]
@orthoxerox Ahh, you mean that the meaning of `return` would be hugely different in the two following cases:


```csharp
string MyMethod()
{
	var name = {
		Log("Getting name");
		return "John";
	}

	return "Dave";
}
```

and

```csharp
string MyMethod()
{
	{
		Log("Getting name");
		return "John";
	}
}
```

**Comment by alrz** [comment]
```
sequence-statement:
    expression-statement
    declaration-statement
```
This syntax would allow a perry narrow set of statements in sequence expressions.

I think we should limit the context in which we can use block expressions to deal with ambiguities instead of limiting statements themselves.

I suggest we allow this in **assignments**, **initializers** and the **switch expression**,
```cs
var x = {
  foreach (var item in list) {
    if (item == value) {
      break true;
    }
  }
  break false;
};


var y = e switch {
  true => {
    using (r) {
      break e.M();
    }
  },
  _ => null
};
```
which is equivalent to:
```cs
int x;
foreach (var item in list) {
  if (item == value) {
    x = true;
    goto _lable;
  }
}
x = false;
_label:

object y;
if (e == true) {
  using (e) {
    y = e.M();
  }
} else {
  y = null;
}
```

To return the result from a block expression we use [`break expr`](http://cr.openjdk.java.net/~briangoetz/amber/switch-rehab.html), `break;` and `return;` statements are permitted and would work the same way that they do today. 


/cc @gafter

**Comment by quinmars** [comment]
What would be the benefit of `if (var x = e; x)` over `var x = e; if (x)`?

**Comment by alrz** [comment]
@quinmars https://github.com/dotnet/csharplang/issues/595

**Comment by quinmars** [comment]
Yes, for while loops this might make sense, but `if`` condintions have leaking scope. So `if (var x = e; x)` and `var x = e; if (x)` would be the same. I don't remeber if `switch` is leaky or not, but I guess it is.

**Comment by alrz** [comment]
This pretty much come down to "inlining a variable declaration" which I don't think could be *that* useful, I prefer "block expressions" because they enable scenarios that is currently not possible with switch expressions,
```cs
var x = e switch {
  true => {
    foreach (var item in list) {
      return; // return from *method* // not possible with local functions
      break; // break the loop
      break true; // terminate and give the result to x
    }
  },
  _ => false
}
```
The [suggested workaround](https://github.com/dotnet/csharplang/issues/1054#issuecomment-340612337) is to use local functions,
```cs
var x = e switch {
  true => LocalFunction(), ...
}
```

`return` in a local function doesn't return from the enclosing method, but the local function itself.

**Comment by qrli** [comment]
@alrz I like your idea. But for the `return`, I think it is confusing. The match syntax is like lambda, but the meaning of `return` would be very different. That's subtle and can be easily overlooked.

I think it should forbid `return` at all, since it is inside an expression.

**Comment by alrz** [comment]
>it should forbid return at all, since it is inside an expression.

See https://github.com/dotnet/csharplang/issues/867 which would allow `return` *as* an expression.

**Comment by gafter** [comment]
I have championed this issue and added the following notes:

UPDATE 2019-09-11: I no longer think there is a need to so severely restrict which statement forms are permitted, though there are now other issues to consider:
- What is the scope of variables declared in these statements, given our existing scoping rules?
- What kinds of statements are permitted?  E.g. we might forbid control transfers out of the expression.
- How does this interact with a tuple expression?

**Comment by orthoxerox** [comment]
I'd say a sequence expression should have its own scope. Variables declared inside it do not leak outside.

Are there any leaky variables beyond out vars that might complicate the decision?

**Comment by CyrusNajmabadi** [comment]
> Are there any leaky variables beyond out vars that might complicate the decision?

Patterns?

**Comment by orthoxerox** [comment]
Oh, right. But I think they have the same solution, since they are leaky for the same purpose.

**Comment by alrz** [comment]
so this is as close as we can get to "block expressions"?

>What kinds of statements are permitted? E.g. we might forbid control transfers out of the expression.

I really wanted to do `var x = e switch { …, _ => { continue; } };` but I guess it wouldn't be possible with sequence expressions. 

>What is the scope of variables declared in these statements, given our existing scoping rules?

I think making the scope to not leak out of itself at all would be too restrictive, I'd suggest we give it the "enclosing scope" which has been originally proposed for patterns and out var
```cs
if ((var x = M(); f(x))) { 
  // x still in scope
}
// but not here
```
```cs
var a = (var x = M(); f(x)); // ok
var b = M(arg, (var y = M(), f(y)), y); // ok
// neither x nor y would not be in scope here
```
Further, if we include *empty-statement* among valid statements, it'd enable to restrict pattern and out var scopes as well.
```cs
if ((;; o is string s)) { } 
// s wouldn't be in scope here because it now follows seq-expression scoping roles!
```
Obviously skipping double parens would definitely be something to consider.

**Comment by Thaina** [comment]
Addition comment

I think Sequence Expressions will be great if we could use it in conjunction with single line `try/catch` block

```C#
var user = try (DoSomething();Loging();await GetUserFromDB(userID)) catch(Expression ex) => (LogException(ex);null);
```

**Comment by CyrusNajmabadi** [comment]
this doesn't seem to be an improvement at all over how this would be written today.

**Comment by Thaina** [comment]
@CyrusNajmabadi Nowaday you can't assign variable out of `try` block. There was already some proposal on `try expression`

I just add that we could also write try expression to assign variable and also write some expression that may throw exception and affect the assignment in the same try block

**Comment by gafter** [comment]
> I think Sequence Expressions will be great if we could use it in conjunction with single line `try/catch` block

That doesn't appear to have anything to do with this proposal.

**Comment by Thaina** [comment]
@gafter I just pointed out that this could be alternative syntax to `try/catch`

Normally we need to defined variable and assign it in the middle. We can't 

```C#
LongNameTypeWithGeneric<LongNameType> obj;
try
{
    DoSomethingThatCanThrowException();
    obj = GetObj();
}
catch(Exception e)
{
    obj = defaultObj;
}
```

It could be simplified if we have both `try expression` and `sequence expression` feature

```C#
var obj = try (DoSomethingThatCanThrowException();GetObj()) catch(Exception e) => defaultObj;
```

**Comment by qrli** [comment]
@Thaina `try-catch` requires braces to disambiguate nesting scenario, even for single statement body.

**Comment by Thaina** [comment]
@qrli `try expression` has its own proposal

**Comment by gafter** [comment]
> this doesn't seem to be an improvement at all over how this would be written today.

@CyrusNajmabadi, how would the following be written today?

```csharp
class C: B
{
    public C(int x) : 
        // shorthand for base(f(x), f(x)) but computing f(x) only once
        base(    (var t = f(x); t),
                 t)
    {
    }
}
```

**Comment by binki** [comment]
@gafter @CyrusNajmabadi 

I think this is what people would have to write today:
```csharp
class C: B
{
    public C(int x) : base(f(x), 0)
    {
    }
    C(int t, short dummy) : base(t, t)
    {
    }
}
```

The shorthand does seem convenient for that example.

**Comment by CyrusNajmabadi** [comment]
I would do what TS does and allow arbitrary code (that does not reference 'this') prior to calling 'base'.  Then you get full language/statement access without having to shoehorn everything into expr/arg-list form. 

The latter feels more like putting the cart before the horse to me.

**Comment by HaloFour** [comment]
Or this:

```cs
class C : B
{
    public C(int x) : base(helper(x, out int y), y) {
    }

    private static int helper(int x, out int returnValue) {
        return returnValue = f(x);
    }
}
```

But I'm with Cyrus, lifting the restriction on calling non-`this` members and coming up with a syntax for invoking the base constructor mid-constructor feels like the better approach for that specific use case.  Cramming sequence expressions into `base` arguments is very difficult to read.

**Comment by gafter** [comment]
And for field initializers (e.g. in a record declaration)?

**Comment by narfanar** [comment]
For the base class constructor case, there's #2335 .

***
***

For the field initializer case:

I can't find the exact example in this issue, so I might be misunderstanding badly.

If it's for record _literals_, then I don't think it looks clearer than having temporaries above the field literal.

Maybe with #2574 fixed we can something a bit clearer (IMHO):
```cs
// (1)
var x = (
    var blah = blah();
    Foo(f(blah), g(blah)
)
// (2)
// instead of
var x = Foo((var blah = blah(); f(blah)), g(blah));
```
This also fixes my issue with having `blah` escape scopes in a weird way (like from one argument's expression to another's) because it then works like normal scopes (because in (1) it's _outside_ both arguments' expressions).

---

If it's for record _definitions_, then:

```cs
data class Foo {
    T0 bar0;
    T1 bar1;
}
// or
data class Foo(T0 bar0, T1 bar1);
// I don't know what the latest design for this looks like...
```

Then maybe (throwing things around):
```cs
data class Foo {
    T0 bar0;
    T1 bar1;
    Foo() {
        var blah = blah();
        bar0 = f(blah);
        bar1 = g(blah);
    }
// or
data class Foo(T0 bar0, T1 bar1) {
    Foo() {
        var blah = blah();
        base(f(blah), g(blah)); // maybe something else besides `base`?
    }
}
```
Here's the catch: Both of those possibilities allow arbitrary code to be run before initialization, even if any particular field will not be using its default value in some instance because it was specified in the literal. But you can do all that with the original proposal, too, otherwise it'd be unsound:
1. If field `a` exposes some name `x` used in the init of field `b` but field `a` is specified in the literal, do you still run `a` default init expression?
2. You can hide any loops in functions called in the expression.

***
***

For the `for` loop case... it's okay.

**Comment by gafter** [comment]
> For the base class constructor case, there's #2335 .

Seems very unlikely to happen, ever.  It is certainly much more complex than sequence expressions.

**Comment by narfanar** [comment]
> Seems very unlikely to happen, ever. It is certainly much more complex than sequence expressions.

That's a bit surprising. It's not proposing any new functionality. Technically just (a bit unconventional) syntax sugar. This would help in this case:

```cs
class C: B
{
    public C(int x)
    : { // Some way to specify you definitely don't mean to run the default `base()`?
        var t = f(x);
        base(t, t); // `base` can only ever be invoked in the main body of a constructor.
    }
}
```

Or, failing that (although I'd really like it...), could we get a spread operator for tuples?

```cs
class C: B
{
    public C(int x)
    : base(*f_for_c(x))
	{}
    (T, T) f_for_c(int x) { var t = f(x); return (t, t); }
}
```

It's just that the constructor example here does not seem like the best pitch for the feature. And honestly I'm already down with anything that gets us closer to clean everything's-an-expression. IMHO, that requires the scope not to leak. But those examples I have issues with (all deal with some form of an ArgumentList) _need_ the scoping hole.

**Comment by a-downing** [comment]
Why couldn't this syntax be used?

```
int x = {
    string str = "123";
    Int32.TryParse(str, out var num) ? num : 0
};
```

If the last expression in the block doesn't end with a semicolon, then the block is a sequence expression. What else could this be ambiguous with? This could also open up the possibility for if statements to become if expressions.

**Comment by CyrusNajmabadi** [comment]
It's not a question of ambiguoty. It's a question of clarity. { Already starts an initializer expression. And having people have to look and see if there's a trailing semicolon or not makes code harder to grok.

**Comment by a-downing** [comment]
Do you mean an anonymous type initializer? But there is no `new` before the `{`, so if you see `{` without `new` where an expression is expected, you automatically know what it is.

**Comment by CyrusNajmabadi** [comment]
No. I mean normal initializer expressions. There used in a few places. Like: `int[] x = { ... }` or `new Foo { X = { ... } }`. Or `new X { { ... } }`. Three are all legal syntaxes today that use `{` to start an expression.

**Comment by a-downing** [comment]
But you still have `new` before `{` in some cases and `,` separating things instead of `;` in all cases.

**Comment by a-downing** [comment]
I guess there is this `int[] arr = {0};` though.

**Comment by CyrusNajmabadi** [comment]
> But you still have new before { in some cases 

No. In none of the cases there is a new before the initializer expression. It's just an expression (where other expressions are allowed). You are proposing a new expression that starts with { that would also be legal in all those contexts. That would be very confusing.

**Comment by 333fred** [comment]
See https://github.com/dotnet/csharplang/issues/3086 for the version of this proposal with braces.

**Comment by Thaina** [comment]
I have proposed https://github.com/dotnet/csharplang/discussions/249 but was moved to discussion
